lib/Rest/VstRequest.cpp

```diff
@@ -72,49 +60,82 @@ VstRequest::VstRequest(ConnectionInfo const& connectionInfo,
 }
 
 VPackSlice VstRequest::payload(VPackOptions const* options) {
-  // message does not need to be validated here, as it was already
-  // validated before
-  return _message.payload();
-}
-
-std::unordered_map<std::string, std::string> const& VstRequest::headers()
-    const {
-  if (!_headers) {
-    using namespace std;
-    _headers = make_unique<unordered_map<string, string>>();
-    VPackSlice meta = _message.header().at(6);  // get meta
-    for (auto const& it : VPackObjectIterator(meta, true)) {
-      // must lower-case the header key
-      _headers->emplace(StringUtils::tolower(it.key.copyString()),
-                        it.value.copyString());
+  TRI_ASSERT(options != nullptr);
+  
+  if (_contentType == ContentType::JSON) {
+    if (!_vpackBuilder) {
+      StringRef json = _message.payload();
+      if (!json.empty()) {
+        _vpackBuilder = VPackParser::fromJson(json.data(), json.length());
+      }
+    }
+    if (_vpackBuilder) {
+      return _vpackBuilder->slice();
+    }
+  } else if (_contentType == ContentType::VPACK) {
+    StringRef vpack = _message.payload();
+    if (!vpack.empty()) {
+      if (!_validatedPayload) {
+        VPackOptions validationOptions = *options; // intentional copy
+        validationOptions.validateUtf8Strings = true;
+        VPackValidator validator(&validationOptions);
+        // will throw on error
+        _validatedPayload = validator.validate(vpack.data(), vpack.length());
+      }
+      return VPackSlice(vpack.data());
     }
   }
-  return *_headers;
-}
-
-std::string const& VstRequest::header(std::string const& key,
-                                      bool& found) const {
-  headers();
-  return lookupStringInMap(*_headers, key, found);
+  return VPackSlice::noneSlice();  // no body
 }
 
-std::string const& VstRequest::header(std::string const& key) const {
-  bool unused = true;
-  return header(key, unused);
+void VstRequest::setHeader(VPackSlice keySlice, VPackSlice valSlice) {
+  if (!keySlice.isString() || !valSlice.isString()) {
+    return;
+  }
+  
+  std::string key = StringUtils::tolower(keySlice.copyString());
+  std::string val = StringUtils::tolower(valSlice.copyString());
+  static const char* mime = "application/json";
+  if (key == StaticStrings::Accept &&
+      val.length() >= 16 && memcmp(val.data(), mime, 16) == 0) {
+    _contentTypeResponse = ContentType::JSON;
+    return;  // don't insert this header!!
+  } else if (key == StaticStrings::ContentTypeHeader &&
+             val.length() >= 16 && memcmp(val.data(), mime, 16) == 0) {
+    _contentType = ContentType::JSON;
+    return;  // don't insert this header!!
+  }
+  
+  // must lower-case the header key
+  _headers.emplace(std::move(key), std::move(val));
```

Comment:
We are now always lower-casing the header *value* here.
I think this is not good. While it is ok to lower-case the header name, the header value should be stored as-is, without lower-casing it.